Classes/common/CDTIncrementalStore.m

```diff
@@ -0,0 +1,2798 @@
+            oops(@"ReadItBack: error: %@", err);
+        }
+        // Always oops
+        if ([upedRev.body count] == 0) oops(@"empty save");
+    }
+
+    return YES;
+}
+
+/**
+ *  Delete a managed object from the database
+ *
+ *  @param mo    Managed Object
+ *  @param error Error
+ *
+ *  @return YES/NO
+ */
+- (BOOL)deleteManagedObject:(NSManagedObject *)mo error:(NSError **)error
+{
+    NSError *err = nil;
+    NSManagedObjectID *moid = [mo objectID];
+    NSString *docID = [self stringReferenceObjectForObjectID:moid];
+
+    /**
+     *  @See CDTISDeleteAggresively
+     */
+    if (CDTISDeleteAggresively) {
+        if (![self.datastore deleteDocumentWithId:docID error:&err]) {
+            if (error) *error = err;
+            return NO;
+        }
+        return YES;
+    }
+
+    CDTDocumentRevision *oldRev = [self.datastore getDocumentWithId:docID error:&err];
+    if (!oldRev) {
+        if (error) *error = err;
+        return NO;
+    }
+
+    if (![self.datastore deleteDocumentFromRevision:oldRev error:&err]) {
+        if (error) *error = err;
+        return NO;
+    }
+    return YES;
+}
+
+/**
+ *  optLock??
+ *
+ *  @param mo    Managed Object
+ *  @param error Error
+ *
+ *  @return YES/NO
+ */
+- (BOOL)optLockManagedObject:(NSManagedObject *)mo error:(NSError **)error
+{
+    oops(@"We don't do this yet");
+    return NO;
+}
+
+/**
+ *  Create a dictionary of values for a the attributes of a Managed Object from
+ *  a docId/ref.
+ *
+ *  @param docID   docID
+ *  @param context context
+ *  @param version version
+ *  @param error   error
+ *
+ *  @return dictionary or nil with error
+ */
+- (NSDictionary *)valuesFromDocID:(NSString *)docID
+                      withContext:(NSManagedObjectContext *)context
+                       versionPtr:(uint64_t *)version
+                            error:(NSError **)error
+{
+    NSError *err = nil;
+
+    CDTDocumentRevision *rev = [self.datastore getDocumentWithId:docID error:&err];
+    if (!rev) {
+        if (error) *error = err;
+        return nil;
+    }
+    NSMutableDictionary *values = [NSMutableDictionary dictionary];
+    for (NSString *name in rev.body) {
+        if ([name isEqualToString:kCDTISObjectVersionKey]) {
+            *version = [rev.body[name] longLongValue];
+            continue;
+        }
+        if ([name hasPrefix:kCDTISPrefix]) {
+            continue;
+        }
+
+        id obj = [self decodeProperty:name
+                              fromDoc:rev.body
+                        withBlobStore:rev.attachments
+                          withContext:context];
+        if (!obj) {
+            // Dictionaries do not take nil, but Values can't have NSNull.
+            // Apparently we just skip it and the properties faults take care
+            // of it
+            continue;
+        }
+        values[name] = obj;
+    }
+
+    return [NSDictionary dictionaryWithDictionary:values];
+}
+
+- (BOOL)commWithRemote:(CDTReplicator *)rep
```

Comment:
It seems that you have embedded the replicator functions into the incremental store.  There is another approach that we should at least consider: exposing the local Cloudant-sync db and allow the application to manage replication.